storage/common/include/algebra/storage/vector.hpp

```diff
@@ -0,0 +1,243 @@
+/** Algebra plugins, part of the ACTS project
+ *
+ * (c) 2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+// Project include(s)
+#include "algebra/qualifiers.hpp"
+#include "algebra/storage/array_operators.hpp"
+
+// System include(s).
+#include <cstddef>
+#include <initializer_list>
+#include <type_traits>
+#include <utility>
+
+namespace algebra::storage {
+
+/// Vector wrapper for AoS vs interleaved SoA data. @c value_t can e.g. be a
+/// SIMD vector.
+template <std::size_t N, typename value_t,
+          template <typename, std::size_t> class array_t>
+class vector {
+ public:
+  // Value type is a simd vector in SoA and a scalar in AoS
+  using value_type = value_t;
+  /// Underlying data array type
+  using array_type = array_t<value_t, N>;
+
+  /// Default contructor sets all entries to zero.
+  constexpr vector() { zero_fill(std::make_index_sequence<N>{}); }
+
+  /// Construct from element values @param vals .
+  ///
+  /// In order to avoid uninitialized values, which deteriorate the performance
+  /// in explicitely vectorized code, the underlying data array is filled with
+  /// zeroes if too few arguments are given.
+  template <typename... Values>
+  constexpr vector(Values &&... vals) : m_data{vals...} {
```

Comment:
@niermann999, I played with this a little. MSVC becomes okay with this line if you give up on the rvalue reference, and go with a simple lvalue one. I.e.

```suggestion
  constexpr vector(const Values &... vals) : m_data{vals...} {
```

I'm anyway not sure why rvalue formalism was chosen here in the first place. :confused:

Reply:
:thinking: Or at least #97 starts working that way... I was doing my tests in that branch, not this one...

Reply:
Great, thanks a lot for having a look! It's a universal reference, so it can be an rvalue or an lvalue. I was hoping to circumvent copies when passing in larger objects, like e.g. `Vc::Vector` or `std::array`

Reply:
Right... I vaguely remember having issues with universal references in MSVC in the past... 🤔 To be honest, the templating here is complicated enough that I can't say at this stage if MSVC is right to complain or not. 😦

I think, if copy and move constructors are important to have here, we would need to declare them separately. 😦

Reply:
Maybe `m_data{std::forward<Values>(vals)...}` will stop MSVC from complaining?

Reply:
I was absolutely convinced that Stephen would be right. But alas, no... That's not the silver bullet after all. 😦

Having played with the code a bit more, the whole shebang gets triggered from line 32 in `benchmark_vector.hpp`

```c++
  std::vector<vector_t> a, b, results;
```

```
C:\Users\krasz\ATLAS\algebra\algebra-plugins\benchmarks\common\include\benchmark\common\benchmark_vector.hpp(32,1): message : see reference to class template instantiation 'std::vector<vector_t,std::allocator<vector_t>>' being compiled [C:\Users\krasz\ATLAS\algebra\build\benchmarks\algebra_benchmark_vc_soa_getter.vcxproj]
          with
          [
              vector_t=algebra::storage::vector<3,Vc_1::Vector<double,Vc_1::VectorAbi::Sse>,algebra::vc_soa::storage_type>
          ]
```

Clearly MSVC wants to use that very generic constructor as the copy/move constructor for the class. And fails at doing so.

Having experimented with a number of different SFINAE ideas on this, trying to only enable this constructor when the arguments are "just so", in the end I gave up. And made these changes instead:

```diff
diff --git a/math/vc_soa/include/algebra/math/impl/vc_soa_vector.hpp b/math/vc_soa/include/algebra/math/impl/vc_soa_vector.hpp
index b47a4a9..b640ab2 100644
--- a/math/vc_soa/include/algebra/math/impl/vc_soa_vector.hpp
+++ b/math/vc_soa/include/algebra/math/impl/vc_soa_vector.hpp
@@ -39,8 +39,8 @@ ALGEBRA_HOST_DEVICE inline storage::vector<N, Vc::Vector<scalar_t>, array_t>
 cross(const storage::vector<N, Vc::Vector<scalar_t>, array_t> &a,
       const storage::vector<N, Vc::Vector<scalar_t>, array_t> &b) {

-  return {Vc::fma(a[1], b[2], -b[1] * a[2]), Vc::fma(a[2], b[0], -b[2] * a[0]),
-          Vc::fma(a[0], b[1], -b[0] * a[1])};
+  return {{Vc::fma(a[1], b[2], -b[1] * a[2]), Vc::fma(a[2], b[0], -b[2] * a[0]),
+           Vc::fma(a[0], b[1], -b[0] * a[1])}};
 }

 /// Dot product between two input vectors
diff --git a/storage/common/include/algebra/storage/vector.hpp b/storage/common/include/algebra/storage/vector.hpp
index 4f0681a..784c3ff 100644
--- a/storage/common/include/algebra/storage/vector.hpp
+++ b/storage/common/include/algebra/storage/vector.hpp
@@ -32,21 +32,10 @@ class vector {
   /// Default contructor sets all entries to zero.
   constexpr vector() { zero_fill(std::make_index_sequence<N>{}); }

-  /// Construct from element values @param vals .
-  ///
-  /// In order to avoid uninitialized values, which deteriorate the performance
-  /// in explicitely vectorized code, the underlying data array is filled with
-  /// zeroes if too few arguments are given.
-  template <typename... Values>
-  constexpr vector(Values &&... vals) : m_data{vals...} {
-    if constexpr ((sizeof...(Values) < N) &&
-                  (!std::conjunction_v<std::is_same<array_type, Values>...>)) {
-      zero_fill(std::make_index_sequence<N - sizeof...(Values)>{});
-    }
-  }
-
   /// Construct from existing array storage @param vals .
   constexpr vector(const array_type &vals) : m_data{vals} {}
+  /// Construct from existing array storage @param vals .
+  constexpr vector(array_type &&vals) : m_data{vals} {}

   /// Assignment operator from wrapped data.
   ///
diff --git a/tests/vc_soa/vc_soa.cpp b/tests/vc_soa/vc_soa.cpp
index d81c421..fbf9030 100644
--- a/tests/vc_soa/vc_soa.cpp
+++ b/tests/vc_soa/vc_soa.cpp
@@ -27,8 +27,8 @@ TEST(test_vc_host, vc_soa_vector) {
   // Value type is Vc::Vector<float>
   using value_t = typename vector3_v::value_type;

-  vector3_v a{value_t::One(), 2.f * value_t::One(), 3.f * value_t::One()};
-  vector3_v b{4.f * value_t::One(), 5.f * value_t::One(), 6.f * value_t::One()};
+  vector3_v a{{value_t::One(), 2.f * value_t::One(), 3.f * value_t::One()}};
+  vector3_v b{{4.f * value_t::One(), 5.f * value_t::One(), 6.f * value_t::One()}};

   EXPECT_TRUE((a[0] == value_t(1.f)).isFull());
   EXPECT_TRUE((a[1] == value_t(2.f)).isFull());
@@ -111,7 +111,7 @@ TEST(test_vc_host, vc_soa_getter) {
   // Value type is Vc::Vector<float>
   using value_t = typename vector3_v::value_type;

-  vector3_v a{value_t::One(), 2.f * value_t::One(), 3.f * value_t::One()};
+  vector3_v a{{value_t::One(), 2.f * value_t::One(), 3.f * value_t::One()}};

   // All results in the vector are the same, so only check the first one
```

I.e. we should just leave it up to the constructor of `std::array` to do its thing, and just put an extra pair of parentheses in the client code where we want to initialize a vector from its elements. 🤔

Reply:
Oh, I saw this post 2 seconds after trying for myself :) Hm, removing this constructor also removes much of the purpose of this class. The point is the call to `zero_fill` (which admittedly does not work right now, because of a Vc deprecation warning that kills the CI), which will correctly initialize a vector register that is larger than the number of passed arguments. I doubt the constructor of `std::array` can do that for us, nor should the client code in detray have to worry about this

Reply:
You could also add `template <typename... Values, std::enable_if_t<std::is_convertible_v<Values, scalar_t> && ..., bool> = true>`.

I'd add the forward either way, even if it's unlikely to matter for most scalar types.

Reply:
A `Vc::Vector` won't be convertible to a scalar (I hope :D), but maybe you are right, the issue with filling the vector register correctly only exists for the AoS case, where the arguments are indeed scalar/integer...

Reply:
Uh yeah sorry I mean test if it's convertible to the relevant `Vc::Vector`. Why wouldn't that work?

Reply:
It could work, just pushed it :)

Reply:
Looks good so far, thanks everyone!